internal/controller/certificates/secrets.go

```diff
@@ -0,0 +1,50 @@
+/*
+Copyright 2020 The cert-manager Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package certificates
+
+import (
+	"crypto/x509"
+	"strings"
+
+	apiutil "github.com/jetstack/cert-manager/pkg/api/util"
+	cmapi "github.com/jetstack/cert-manager/pkg/apis/certmanager/v1"
+	utilpki "github.com/jetstack/cert-manager/pkg/util/pki"
+)
+
+// AnnotationsForCertificateSecret returns a map which is set on all
+// Certificate Secret's Annotations when issued. These annotations contain
+// information about the Issuer and Certificate.
+// If the X.509 certificate is not-nil, additional annotations will be added
+// relating to its Common Name and Subject Alternative Names.
+func AnnotationsForCertificateSecret(crt *cmapi.Certificate, certificate *x509.Certificate) map[string]string {
+	annotations := make(map[string]string)
+
+	annotations[cmapi.CertificateNameKey] = crt.Name
+	annotations[cmapi.IssuerNameAnnotationKey] = crt.Spec.IssuerRef.Name
+	annotations[cmapi.IssuerKindAnnotationKey] = apiutil.IssuerKind(crt.Spec.IssuerRef)
+	annotations[cmapi.IssuerGroupAnnotationKey] = crt.Spec.IssuerRef.Group
```

Comment:
Note (not really related to this PR): I went to our CI cluster to look for an example cert to see if we always added _all_ these annotations and noticed that at least this `Certificate`'s `Secret` had a duplicate set of `certmanager.io`, `cert-manager.k8s.io` annotations:
```
metadata:
  annotations:
    cert-manager.io/alt-names: prow.build-infra.jetstack.net
    cert-manager.io/certificate-name: prow-tls
    cert-manager.io/common-name: prow.build-infra.jetstack.net
    cert-manager.io/ip-sans: ""
    cert-manager.io/issuer-group: cert-manager.io
    cert-manager.io/issuer-kind: ClusterIssuer
    cert-manager.io/issuer-name: letsencrypt-prod
    cert-manager.io/uri-sans: ""
    certmanager.k8s.io/alt-names: prow.build-infra.jetstack.net
    certmanager.k8s.io/common-name: prow.build-infra.jetstack.net
    certmanager.k8s.io/ip-sans: ""
    certmanager.k8s.io/issuer-kind: ClusterIssuer
    certmanager.k8s.io/issuer-name: letsencrypt-prod
  creationTimestamp: "2019-08-30T08:59:45Z"
```
There is I guess nothing in that that would affect this PR though, was just curious how it ended up with a duplicate set.